File: 백준/Gold/1937. 욕심쟁이 판다/욕심쟁이 판다.cc
#include <iostream>
#include <algorithm>
#include <vector>
#include <cstring>

using namespace std;

int n;
int graph[500][500];
int dp[500][500];
int dx[4] = {1, -1, 0, 0};
int dy[4] = {0, 0, 1, -1};

int dfs(int x, int y) {
    // 이미 계산된 경우 바로 반환
    if (dp[y][x] != 0) return dp[y][x];

    // 현재 위치는 최소 1칸 방문 가능
    dp[y][x] = 1;

    for (int i = 0; i < 4; i++) {
        int nx = x + dx[i];
        int ny = y + dy[i];

        // 범위 체크 및 대나무 양 조건 확인
        if (nx >= 0 && nx < n && ny >= 0 && ny < n && graph[y][x] < graph[ny][nx]) {
            dp[y][x] = max(dp[y][x], dfs(nx, ny) + 1);
        }
    }

    return dp[y][x];
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    cin >> n;
    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++) cin >> graph[i][j];
    
    // 메모이제이션 배열 초기화
    memset(dp, 0, sizeof(dp));

    int answer = 0;
    // 모든 지점에서 시작해서 최대 이동 칸 수 찾기
    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++) 
            answer = max(answer, dfs(j, i));

    cout << answer;

    return 0;
}

File: 백준/Gold/1937. 욕심쟁이 판다/욕심쟁이 판다.java
import java.util.*;
import java.io.*;

public class Main {
    static int n;
    static int[] dx = {0, 1, 0, -1}, dy = {-1, 0, 1, 0};
    static int[][] graph;
    static long answer = 0;
    static long[][] dp;

    static boolean inRange(int x, int y) {
        return (-1 < x && x < n && -1 < y && y < n);
    }

    static long dfs(int x, int y) {
        if (dp[y][x] != 1) return dp[y][x];

        long result = 1;
        for (int i = 0; i < 4; i++) {
            int nx = x + dx[i];
            int ny = y + dy[i];

            if (!inRange(nx, ny)) continue;

            if (graph[y][x] < graph[ny][nx]) result = Math.max(result, 1 + dfs(nx, ny));
        }

        return dp[y][x] = result;
    }
    
    // DFS + DP
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        n = Integer.parseInt(br.readLine());
        graph = new int[n][n];
        dp = new long[n][n];

        for (int i = 0; i < n; i++) {
            Arrays.fill(dp[i], 1L);
            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int j = 0; j < n; j++) graph[i][j] = Integer.parseInt(st.nextToken());
        }

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                dp[i][j] = dfs(j, i);
                if (dp[i][j] > answer) answer = dp[i][j];
            }
        }

        bw.write(answer + "");
        bw.flush();
        bw.close();
        br.close();
    }
}